File: Trapping Rain Water - Leetcode 42.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l_wall = r_wall = 0
        n = len(height)
        max_left = [0] * n
        max_right = [0] * n

        for i in range(n):
            j = -i - 1
            max_left[i] = l_wall
            max_right[j] = r_wall
            l_wall = max(l_wall, height[i])
            r_wall = max(r_wall, height[j])

        summ = 0
        for i in range(n):
            pot = min(max_left[i], max_right[i])
            summ += max(0, pot - height[i])

        return summ

# Time Complexity: O(n)
# Space Complexity: O(n)

File: Trapping Rain Water - Leetcode 42/Trapping Rain Water - Leetcode 42.js
var trap = function(height) {
    const n = height.length;
    const maxLeft = Array(n).fill(0);
    const maxRight = Array(n).fill(0);
    
    let lWall = 0, rWall = 0;
    
    for (let i = 0; i < n; i++) {
        const j = n - i - 1;
        maxLeft[i] = lWall;
        maxRight[j] = rWall;
        lWall = Math.max(lWall, height[i]);
        rWall = Math.max(rWall, height[j]);
    }
    
    let sum = 0;
    for (let i = 0; i < n; i++) {
        const pot = Math.min(maxLeft[i], maxRight[i]);
        sum += Math.max(0, pot - height[i]);
    }
    
    return sum;
};
